applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp

```diff
@@ -0,0 +1,238 @@
+        new_prop->SetId(++unique_prop_id);
+        mrDestinationModelPart.AddProperties(new_prop);
+        replicated_properties[it_prop->Id()] = new_prop;
+    }
+
+    // Note: we need to manually create the geometry with the cloned nodes
+    for (IndexType i = 0; i < mrOriginModelPart.NumberOfElements(); ++i)
+    {
+        auto it_elem = mrOriginModelPart.ElementsBegin() + i;
+        PointerVector<Node<3>> new_elem_nodes;
+        for (auto& node : it_elem->GetGeometry())
+        {
+            new_elem_nodes.push_back(mReplicatedNodes[node.Id()]);
+        }
+        auto new_elem = it_elem->Clone(++unique_elem_id, new_elem_nodes);
+        new_elem->SetProperties(replicated_properties[it_elem->GetProperties().Id()]);
+        mrDestinationModelPart.AddElement(new_elem);
+        if (mReplicateSubModelParts)
+        {
+            auto tag = origin_elements_tags[unique_elem_id];
+            dest_elems_of_collection[tag].push_back(unique_elem_id);
+        }
+    }
+
+    // Note: we need to manually create the geometry with the cloned nodes
+    for (IndexType i = 0; i < mrOriginModelPart.NumberOfConditions(); i++)
+    {
+        auto it_cond = mrOriginModelPart.ConditionsBegin() + i;
+        PointerVector<Node<3>> new_cond_nodes;
+        for (auto& node : it_cond->GetGeometry())
+        {
+            new_cond_nodes.push_back(mReplicatedNodes[node.Id()]);
+        }
+        auto new_cond = it_cond->Clone(++unique_cond_id, new_cond_nodes);
+        new_cond->SetProperties(replicated_properties[it_cond->GetProperties().Id()]);
+        mrDestinationModelPart.AddCondition(new_cond);
+        if (mReplicateSubModelParts)
+        {
+            auto tag = origin_conditions_tags[unique_cond_id];
+            dest_conds_of_collection[tag].push_back(unique_cond_id);
+        }
+    }
+
+    // We need to replicate the sub model parts and then, populate them
+    if (mReplicateSubModelParts)
+    {
+        auto model_part_names = mrOriginModelPart.GetSubModelPartNames();
+        for (auto& name : model_part_names)
+        {
+            mrDestinationModelPart.CreateSubModelPart(name);
+        }
+        
+        // Finally, we add the nodes, elements and conditions to the sub model parts
+        for (auto model_part_collection : collections)
+        {
+            IndexType tag = model_part_collection.first;
+            if (tag != 0) // Tag = 0 means the root model part
+            {
+                for (auto name : model_part_collection.second)
+                {
+                    ModelPart& model_part = mrDestinationModelPart.GetSubModelPart(name);
+                    model_part.AddNodes(dest_nodes_of_collection[tag]);
+                    model_part.AddElements(dest_elems_of_collection[tag]);
+                    model_part.AddConditions(dest_conds_of_collection[tag]);
+                }
+            }
+        }
+    } // if (mReplicateSubModelParts)
+}
+
+
+void ReplicateModelPartUtility::SetOriginMeshPosition()
```

Comment:
The method name is misleading. I'd suggest to rename it to `InitializeOriginMeshZCoordinate`.

Reply:
Method renamed